src/Feldspar/Onnx/onnxToFeld.hs

```diff
@@ -99,49 +104,49 @@ buildValues TD.DOUBLE     t = D.foldMap (\ x -> B.doubleDec x      <> B.string8
 buildValues TD.INT32      t = D.foldMap (\ x -> B.int32Dec x       <> B.string8 "\n") $ TP.int32_data t
 buildValues TD.UINT64     t = D.foldMap (\ x -> B.word64Dec x      <> B.string8 "\n") $ TP.uint64_data t
 buildValues TD.INT64      t = D.foldMap (\ x -> B.int64Dec x       <> B.string8 "\n") $ TP.int64_data t
-buildValues TD.STRING     t = D.foldMap (\ x -> B.lazyByteString x     <> B.string8 "\n") $ TP.string_data t
-buildValues td            _ = error $ "onnxToFeld.buildValues: unsupported element type " ++ showElemT td
+buildValues TD.STRING     t = D.foldMap (\ x -> B.lazyByteString x <> B.string8 "\n") $ TP.string_data t
+buildValues td            _ = error $ "onnxToFeld.buildValues: unsupported element type " ++ U.toString (showElemT td)
 
 -- | Construct a Feldspar program corresponding to the ONNX graph
-mkProgramFile :: G.GraphProto -> String
+mkProgramFile :: G.GraphProto -> L.ByteString
 mkProgramFile gr 
-  = unlines [ "module Main where"
+  = L.unlines [ "module Main where"
             , ""
             , "import Feldspar"
             , "import Feldspar.Compiler (program)"
             , "import Feldspar.Onnx.Operators"
             , ""
-            , "main = program " ++ name
+            , "main = program " <> name
             , ""
-            , name ++ " " ++ unwords params ++ " = " ++ tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
+            , name <> " " <> L.unwords params <> " = " <> tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
             , "  where "
-              ++ intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
+              <> L.intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
```

Comment:
Can we avoid the list conversion by using `D.concatMap`?

Reply:
I'm not sure I follow you here. L.intercalate really wants a list of ByteStrings.

Reply:
Replace `concatMap` with `D.concatMap` and remove `D.toList` from the definition of `nodes`?

Reply:
Sure. I thought Data.Foldable.concatMap would use the foldable type for the resulting list, not just for the input list.